{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "class CountHaplotype:\n",
    "    def __init__(self, masterdata: list) -> None:\n",
    "        self.masterdata = masterdata\n",
    "        \n",
    "    def count_haplotype(self):\n",
    "        result_list = []\n",
    "\n",
    "        for sample in self.masterdata:\n",
    "            gender = sample['gender']\n",
    "            for gene_entry in sample.get('gene_detail', []):  # Use get to handle missing 'gene_detail'\n",
    "                gene = gene_entry.get('gene', 'Unknown')  # Use get to handle missing 'gene'\n",
    "                haplotype_list = gene_entry.get('haplotype', [])\n",
    "\n",
    "                number_allele = len(haplotype_list)\n",
    "\n",
    "                count_dict = {}\n",
    "                \n",
    "                for allele in haplotype_list:\n",
    "                    if number_allele <= 2:\n",
    "                        count_dict[allele] = count_dict.get(allele, 0) + (1 * (2 / number_allele))\n",
    "                    else:\n",
    "                        count_dict[allele] = np.nan\n",
    "\n",
    "                for key, value in count_dict.items():\n",
    "                    result_list.append({'gene': gene, 'gender': gender, 'haplotype': key, 'haplotype_count': value})\n",
    "\n",
    "        result_df = pd.DataFrame(result_list).dropna()\n",
    "        \n",
    "        # Group and aggregate the results using Pandas\n",
    "        result_all = (\n",
    "            pd.DataFrame(result_list)\n",
    "            .dropna()\n",
    "            .groupby(['gene', 'gender', 'haplotype'])\n",
    "            .sum()\n",
    "            .reset_index()\n",
    "            .sort_values(by=\"gene\")\n",
    "            .to_dict(orient='records')\n",
    "        )\n",
    "        \n",
    "        result_male = result_df[result_df['gender'] == 'male']\n",
    "        result_female = result_df[result_df['gender'] == 'female']\n",
    "        \n",
    "        result_by_gender = {\n",
    "            'all': result_all,\n",
    "            'male': result_male.to_dict(orient='records'),\n",
    "            'female': result_female.to_dict(orient='records')\n",
    "        }\n",
    "        \n",
    "        return result_by_gender"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('.out/master_data.json', 'r') as file:\n",
    "        existing_data = json.load(file)\n",
    "\n",
    "obj = CountHaplotype(existing_data)\n",
    "\n",
    "def tran():\n",
    "    data = obj.count_haplotype()\n",
    "\n",
    "    df_all = pd.DataFrame(data['all'])\n",
    "    df_all['gender'] = 'all'\n",
    "    df_all['countkey'] = df_all['gender']+'-'+df_all['gene']+'-'+df_all['haplotype']\n",
    "    df_all['totalkey'] = df_all['gender']+'-'+df_all['gene']\n",
    "    df_all['lookupkey'] = df_all['gene']+'-'+df_all['haplotype']\n",
    "\n",
    "    df_male = pd.DataFrame(data['male'])\n",
    "    df_male['lookupkey'] = 'male'\n",
    "    df_male['countkey'] = df_male['gender']+'-'+df_male['gene']+'-'+df_male['haplotype']\n",
    "    df_male['totalkey'] = df_male['gender']+'-'+df_male['gene']\n",
    "    df_male['lookupkey'] = df_male['gene']+'-'+df_male['haplotype']\n",
    "\n",
    "    df_female = pd.DataFrame(data['female'])\n",
    "    df_female['lookupkey'] = 'female'\n",
    "    df_female['countkey'] = df_female['gender']+'-'+df_female['gene']+'-'+df_female['haplotype']\n",
    "    df_female['totalkey'] = df_female['gender']+'-'+df_female['gene']\n",
    "    df_female['lookupkey'] = df_female['gene']+'-'+df_female['haplotype']\n",
    "\n",
    "    df = pd.concat([df_all, df_male, df_female], ignore_index=True)[['gender','countkey', 'totalkey','lookupkey', 'haplotype_count']].rename(columns={'gender': 'cou_gender'})\n",
    "\n",
    "    df_count_hap = df.groupby('countkey', as_index=False)['haplotype_count'].sum().rename(columns={'haplotype_count': 'hap_cou'})\n",
    "\n",
    "    df_count_total = df.groupby('totalkey', as_index=False)['haplotype_count'].sum().rename(columns={'haplotype_count': 'hap_total'})\n",
    "\n",
    "    df_merge_cou = pd.merge(df, df_count_hap, how='inner', on='countkey')[['countkey','cou_gender', 'totalkey','lookupkey', 'hap_cou']].drop_duplicates().sort_values(by='lookupkey')\n",
    "    df_merge_cou_total = pd.merge(df_merge_cou, df_count_total, how='inner', on='totalkey')[['countkey','cou_gender', 'totalkey','lookupkey', 'hap_cou', 'hap_total']].drop_duplicates().sort_values(by='lookupkey')\n",
    "\n",
    "    master_data_path = '.out/master_data.json'\n",
    "\n",
    "    result = []\n",
    "\n",
    "    with open(master_data_path, 'r') as file:\n",
    "        master_data = json.load(file)\n",
    "        \n",
    "    for i in master_data:   \n",
    "        for j in i['gene_detail']:\n",
    "            for k in j['haplotype']:\n",
    "                master_result = {}\n",
    "                master_result['sample_id'] = i['sample_id']\n",
    "                master_result['nbt_id'] = i['nbt_id']\n",
    "                master_result['gender'] = i['gender']\n",
    "                master_result['gene'] = j['gene']\n",
    "                master_result['haplotype'] = k\n",
    "                master_result['lookupkey'] = j['gene']+'-'+k\n",
    "                result.append(master_result)\n",
    "                \n",
    "    df_master = pd.DataFrame(result)\n",
    "\n",
    "    df_merge = pd.merge(df_master, df_merge_cou_total, how='inner', on='lookupkey').sort_values(by=['sample_id', 'gene'])\n",
    "    df_merge['hap_ratio'] = (df_merge['hap_cou']/df_merge['hap_total']).round(4)\n",
    "    df_merge.fillna(0, inplace=True)\n",
    "\n",
    "    pgx_count_haplotype = []\n",
    "\n",
    "    # Get unique combinations of 'sample_id' and 'gene'\n",
    "    unique_combinations = df_merge[['sample_id', 'gene']].drop_duplicates()\n",
    "\n",
    "    for index, row in unique_combinations.iterrows():\n",
    "        sample_id = row['sample_id']\n",
    "        gene = row['gene']\n",
    "        \n",
    "        # Filter the DataFrame for the current 'sample_id' and 'gene' combination\n",
    "        filtered_df = df_merge[(df_merge['sample_id'] == sample_id) & (df_merge['gene'] == gene)]\n",
    "        \n",
    "        # Get unique 'haplotype' values for the current combination\n",
    "        unique_haplotype = filtered_df['haplotype'].unique()\n",
    "        \n",
    "        for hap in unique_haplotype:\n",
    "            dict_result = {\n",
    "                'cou_hap_sample_id': sample_id,\n",
    "                'cou_hap_nbt_id': sample_id,  # Assuming this should also be 'sample_id'\n",
    "                'cou_hap_gene': gene,\n",
    "                'cou_hap_haplotype': hap,\n",
    "                'cou_hap_gender_distribution': []\n",
    "            }\n",
    "            \n",
    "            # Filter the DataFrame for the current 'sample_id', 'gene', and 'haplotype' combination\n",
    "            filtered_hap_df = filtered_df[filtered_df['haplotype'] == hap]\n",
    "            \n",
    "            for index, hap_row in filtered_hap_df.iterrows():\n",
    "                gender_distribution = {\n",
    "                    'cou_gender': hap_row['cou_gender'],\n",
    "                    'hap_cou': hap_row['hap_cou'],\n",
    "                    'hap_total': hap_row['hap_total'],\n",
    "                    'hap_ratio': hap_row['hap_ratio']\n",
    "                }\n",
    "                dict_result['cou_hap_gender_distribution'].append(gender_distribution)\n",
    "            \n",
    "            pgx_count_haplotype.append(dict_result)\n",
    "    return pgx_count_haplotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "CountHaplotype.count_haplotype() takes 1 positional argument but 2 were given",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[29], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mpprint\u001b[39;00m \u001b[39mimport\u001b[39;00m pprint\n\u001b[0;32m----> 2\u001b[0m pprint(tran())\n",
      "Cell \u001b[0;32mIn[28], line 8\u001b[0m, in \u001b[0;36mtran\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mtran\u001b[39m():\n\u001b[0;32m----> 8\u001b[0m     data \u001b[39m=\u001b[39m obj\u001b[39m.\u001b[39;49mcount_haplotype(existing_data)\n\u001b[1;32m     10\u001b[0m     df_all \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame(data[\u001b[39m'\u001b[39m\u001b[39mall\u001b[39m\u001b[39m'\u001b[39m])\n\u001b[1;32m     11\u001b[0m     df_all[\u001b[39m'\u001b[39m\u001b[39mgender\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39mall\u001b[39m\u001b[39m'\u001b[39m\n",
      "\u001b[0;31mTypeError\u001b[0m: CountHaplotype.count_haplotype() takes 1 positional argument but 2 were given"
     ]
    }
   ],
   "source": [
    "from pprint import pprint\n",
    "pprint(tran())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "710c01454f3e205cd665fbe0fd95fe8b49b32872a549ad1bcc7c06dd7d853108"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
